batch/batch/driver/instance_collection/pool.py

```diff
@@ -312,47 +305,55 @@ async def ready_cores_mcpu_from_estimated_job_queue(self):
         for user, share in user_share.items():
             user_job_query = f'''
 (
-SELECT CAST(COALESCE(SUM(cores_mcpu), 0) AS SIGNED) AS ready_cores_mcpu
-FROM (
-  (
-    SELECT cores_mcpu
-    FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_cancelled)
-    LEFT JOIN batches ON jobs.batch_id = batches.id
-    WHERE user = %s AND batches.`state` = 'running' AND jobs.state = 'Ready' AND always_run = 1 AND inst_coll = %s
-    GROUP BY jobs.batch_id, jobs.job_id
-    ORDER BY jobs.batch_id ASC, jobs.job_id ASC
-    LIMIT {share * JOB_QUEUE_SCHEDULING_WINDOW_SECONDS}
-  )
-  UNION (
-    SELECT cores_mcpu
-    FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_cancelled)
-    LEFT JOIN batches ON jobs.batch_id = batches.id
-    LEFT JOIN batches_cancelled ON batches.id = batches_cancelled.id
-    WHERE user = %s AND batches.`state` = 'running' AND jobs.state = 'Ready' AND always_run = 0 AND batches_cancelled.id IS NULL AND inst_coll = %s
-    GROUP BY jobs.batch_id, jobs.job_id
-    ORDER BY jobs.batch_id ASC, jobs.job_id ASC
-    LIMIT {share * JOB_QUEUE_SCHEDULING_WINDOW_SECONDS}
-  )
-) AS t
-LIMIT {share * JOB_QUEUE_SCHEDULING_WINDOW_SECONDS}
+  SELECT user, jobs.batch_id, jobs.job_id, cores_mcpu,
+    JSON_ARRAYAGG(region) AS regions,
+    COUNT(region) AS n_regions,
+    ROW_NUMBER() OVER (ORDER BY batch_id ASC, job_id ASC) DIV {share} AS scheduling_iteration
+  FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_cancelled)
+  LEFT JOIN batches ON jobs.batch_id = batches.id
+  LEFT JOIN batches_cancelled ON batches.id = batches_cancelled.id
+  LEFT JOIN job_regions ON jobs.batch_id = job_regions.batch_id AND jobs.job_id = job_regions.job_id
+  LEFT JOIN region_ids ON job_regions.region_id = region_ids.region_id
+  WHERE user = %s AND batches.`state` = 'running' AND jobs.state = 'Ready' AND (always_run = 1 OR always_run = 0 AND batches_cancelled.id IS NULL) AND inst_coll = %s
```

Comment:
@danking Can you verify this query is okay? I'm not sure why Cotton had this query split into two before (one for always_run and the other for not always_run and not cancelled). The OR in the where statement is different from before: `(always_run = 1 OR always_run = 0 AND batches_cancelled.id IS NULL)`

Query plan looks like this on the batch database:

```
mysql>   EXPLAIN SELECT user, jobs.batch_id, jobs.job_id, cores_mcpu,
    ->     ROW_NUMBER() OVER (ORDER BY batch_id ASC, job_id ASC) DIV 300 AS scheduling_iteration
    ->   FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_cancelled)
    ->   LEFT JOIN batches ON jobs.batch_id = batches.id
    ->   LEFT JOIN batches_cancelled ON batches.id = batches_cancelled.id
    ->   WHERE user = 'ci' AND batches.`state` = 'running' AND jobs.state = 'Ready' AND (always_run = 1 OR always_run = 0 AND batches_cancelled.id IS NULL) AND inst_coll = 'standard'
    ->   GROUP BY jobs.batch_id, jobs.job_id
    ->   ORDER BY jobs.batch_id ASC, jobs.job_id ASC
    ->   LIMIT 300;
+----+-------------+-------------------+------------+--------+------------------------------------------+------------------------------------------+---------+------------------------+------+----------+-----------------------------------------------------------+
| id | select_type | table             | partitions | type   | possible_keys                            | key                                      | key_len | ref                    | rows | filtered | Extra                                                     |
+----+-------------+-------------------+------------+--------+------------------------------------------+------------------------------------------+---------+------------------------+------+----------+-----------------------------------------------------------+
|  1 | SIMPLE      | batches           | NULL       | ref    | PRIMARY,batches_user_state               | batches_user_state                       | 424     | const,const            |    1 |   100.00 | Using where; Using index; Using temporary; Using filesort |
|  1 | SIMPLE      | batches_cancelled | NULL       | eq_ref | PRIMARY                                  | PRIMARY                                  | 8       | batch.batches.id       |    1 |   100.00 | Using index                                               |
|  1 | SIMPLE      | jobs              | NULL       | ref    | jobs_batch_id_state_always_run_cancelled | jobs_batch_id_state_always_run_cancelled | 130     | batch.batches.id,const |   20 |     1.90 | Using index condition; Using where                        |
+----+-------------+-------------------+------------+--------+------------------------------------------+------------------------------------------+---------+------------------------+------+----------+-----------------------------------------------------------+
3 rows in set, 2 warnings (0.00 sec)
```

Reply:
OK in what sense? Performance?

Reply:
Yes. Not sure why it was split to begin with.

Reply:
Actually, just realized I have a bug in that code. Let me fix it.

Reply:
```
mysql> EXPLAIN SELECT user, jobs.batch_id, jobs.job_id, cores_mcpu,
    ->     ROW_NUMBER() OVER (ORDER BY batch_id ASC, job_id ASC) DIV 300 AS scheduling_iteration
    ->   FROM jobs FORCE INDEX(jobs_batch_id_state_always_run_cancelled)
    ->   LEFT JOIN batches ON jobs.batch_id = batches.id
    ->   LEFT JOIN batches_cancelled ON batches.id = batches_cancelled.id
    ->   WHERE user = 'ci' AND batches.`state` = 'running' AND jobs.state = 'Ready' AND (always_run = 1 OR (always_run = 0 AND batches_cancelled.id IS NULL)) AND inst_coll = 'standard'
    ->   GROUP BY jobs.batch_id, jobs.job_id
    ->   ORDER BY jobs.batch_id ASC, jobs.job_id ASC
    ->   LIMIT 300;
+----+-------------+-------------------+------------+--------+------------------------------------------+------------------------------------------+---------+------------------------+------+----------+-----------------------------------------------------------+
| id | select_type | table             | partitions | type   | possible_keys                            | key                                      | key_len | ref                    | rows | filtered | Extra                                                     |
+----+-------------+-------------------+------------+--------+------------------------------------------+------------------------------------------+---------+------------------------+------+----------+-----------------------------------------------------------+
|  1 | SIMPLE      | batches           | NULL       | ref    | PRIMARY,batches_user_state               | batches_user_state                       | 424     | const,const            |    2 |   100.00 | Using where; Using index; Using temporary; Using filesort |
|  1 | SIMPLE      | batches_cancelled | NULL       | eq_ref | PRIMARY                                  | PRIMARY                                  | 8       | batch.batches.id       |    1 |   100.00 | Using index                                               |
|  1 | SIMPLE      | jobs              | NULL       | ref    | jobs_batch_id_state_always_run_cancelled | jobs_batch_id_state_always_run_cancelled | 130     | batch.batches.id,const |   20 |     1.90 | Using index condition; Using where                        |
+----+-------------+-------------------+------------+--------+------------------------------------------+------------------------------------------+---------+------------------------+------+----------+-----------------------------------------------------------+
3 rows in set, 2 warnings (0.01 sec)
```

Reply:
I don't think we need `always_run = 1`. Also, let's switch this away from `= 0` and `= 1`, I find that more confusing than treating this field as a boolean.

This query, AFAICT, is trying to find jobs that are ready to be scheduled. That's basically:

```
job.state = 'Ready' AND (always_run OR batches_cancelled.id IS NULL)
```

Which reads as: "is the job ready and either its always run or the batch is not cancelled?"

Reply:
Does it matter that the index has `always_run` in it?